src/main/ruby/core/type.rb

```diff
@@ -644,6 +644,18 @@ def self.rb_inspect(val)
       end
     end
 
+    def self.rb_obj_as_string(obj)
+      return obj if object_kind_of?(obj, String)
+      str = obj.to_s
+      return Rubinius::Type.rb_any_to_s(obj) unless object_kind_of?(str, String)
+      Rubinius::Type.infect(str, obj)
```

Comment:
Should the infect only apply is `str` is already a String?

Reply:
The `ToSNode` called by `rb_any_to_s` also infects.